### PR TITLE
eslint-plugin-react-hooks: Add support for ESLint v9

### PR DESCRIPTION
**Pull request by eps1lon**
Based on https://github.com/facebook/react/pull/28774

Does not add support for flat configs i.e. does not adress https://github.com/facebook/react/issues/28313. 

The PR only adds support for ESLint v9 with the old config format and sets up testing infra for  separate ESLint versions. The only breaking changes that affected the plugin was around removal of `context.getScope` and `context.getSource` which have simple replacements in V9 so that we don't have to fork too much to be able to support ESLint V9 and earlier.

The last commit makes it so that we have no breaking changes. If we decide to drop support for ESLint < 9.x, we can just revert [`80e7b49` (#28773)](https://github.com/facebook/react/pull/28773/commits/80e7b49254a14c0799efc573930d9cb108498630)

Alternate to https://github.com/facebook/react/pull/28772

### CODE REVIEW

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js [hunk 1]
```diff
@@ -67,6 +67,27 @@ export default {
...
+     */
+    function getScope(node) {
+      if (typeof context.getScope === 'function') {
+        return context.getScope();
+      } else {
+        return context.sourceCode.getScope(node);
+      }
+    }
+
     const scopeManager = context.getSourceCode().scopeManager;
```
Comment (Debbl) [inline packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:91]: https://eslint.org/blog/2023/09/preparing-custom-rules-eslint-v9/#context-methods-becoming-properties
I noticed that this `context.getSourceCode()` was replaced with `context.sourceCode`, you might want to confirm that.
```js
const sourceCode = context.sourceCode ?? context.getSourceCode();
const scopeManager = sourceCode.scopeManager;
```

Comment (eps1lon) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:91]: `getSourceCode` was only deprecated.

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js [hunk 2]
```diff
@@ -8250,49 +8272,119 @@ describe('react-hooks', () => {
     invalid: [...testsTypescript.invalid, ...tests.invalid],
   };
 
-  new ESLintTester({
+  new ESLintTesterV7({
     parser: require.resolve('@typescript-eslint/parser-v2'),
-    parserOptions,
+    parserOptions: parserOptionsV7,
+  }).run(
+    'eslint: v9, parser: @typescript-eslint/parser@2.x',
```
Comment (poteto) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:8279]: Is this a typo?

```suggestion
    'eslint: v7, parser: @typescript-eslint/parser@2.x',
```

**Review by poteto — APPROVED** [review]

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js (hunk 2 above)
Comment (eps1lon) [reply packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:8279]: good catch, c&p mistakes

File: yarn.lock
[the remark this answers is not in the document]
Comment (eps1lon) [reply yarn.lock:1]: It kind of doesn't make sense for this commit to stay if we revert ESlint v9 support. And if we need to revert this, we also need to revert ESLint v9 support.

I rebased this PR to isolate these changes so if we do need to bisect later, we can still do it on this branch,

[the remark this answers is not in the document]
Comment (poteto) [reply yarn.lock:1]: ah right, sorry I missed that there was a package.json update

**Comment by silkfire** [comment]
When will this be released?

**Comment by RomkeSandwave** [comment]
Still getting `context.getSource is not a function` in 4.6.2. Has this not been released yet?

**Comment by eps1lon** [comment]
> Still getting context.getSource is not a function in 4.6.2. Has this not been released yet?

This PR is released in `eslint-plugin-react-hooks@canary`. Stable will be released soon,

**Comment by dzcpy** [comment]
So how to config it?

**Comment by thenbe** [comment]
@dzcpy `npm install -D eslint-plugin-react-hooks@canary`

**Comment by ArnaudBarre** [comment]
@eps1lon Can we have a stable release of the eslint plugin without waiting for React 19?
With plugin-react now supporting ESLint 9, it would be great to ship #12860 to new Vite users without using compat layer or a canary version.

**Comment by lllapland** [comment]
> So how to config/use it?

yarn add eslint-plugin-react-hooks@5.1.0-rc-eb3ad065-20240822 -D

**Comment by rburgst** [comment]
when can we expect a release?

**Comment by xuhdev** [comment]
Looks like we can install `5.1.0-beta-26f2496093-20240514` for now...

**Comment by c10b10** [comment]
For who made this work:

I'm using the rc version of 5.1.0, but still getting the same error when importing this like this

```
import { fixupConfigRules } from "@eslint/compat";
import eslint from "@eslint/js";
import eslintImportX from "eslint-plugin-import-x";
// import eslintPluginReactHooks from "eslint-plugin-react-hooks";
import eslintReactJsxRuntime from "eslint-plugin-react/configs/jsx-runtime.js";
import eslintReactRecommended from "eslint-plugin-react/configs/recommended.js";
import globals from "globals";
import tsEslint from "typescript-eslint";
 
const typescriptReactApp = [
  { languageOptions: { globals: globals.browser } },
  eslint.configs.recommended,
  ...tsEslint.configs.recommended,
  eslintPluginReactHooks,
  ...fixupConfigRules(eslintReactRecommended),
  ...fixupConfigRules(eslintReactJsxRuntime),
  {
    rules: {
      "react/prop-types": "off",
    },
    files: ["**/*.{js,mjs,cjs,ts,jsx,tsx}"]
  },
];
```

**Comment by EmilYanbaev** [comment]
My working config
 "eslint-plugin-react-hooks": "5.1.0-beta-26f2496093-20240514"
```
import globals from 'globals'
import pluginJs from '@eslint/js'
import tseslint from 'typescript-eslint'
import pluginReact from 'eslint-plugin-react'
import eslintPluginPrettierRecommended from 'eslint-plugin-prettier/recommended'
import pluginJest from 'eslint-plugin-jest'
import pluginReactHooks from 'eslint-plugin-react-hooks'

export default [
  { languageOptions: { globals: globals.browser } },
  pluginJs.configs.recommended,
  ...tseslint.configs.recommended,
  pluginReact.configs.flat.recommended,
  pluginJest.configs['flat/recommended'],
  eslintPluginPrettierRecommended,
  {
    files: ['src/**/*.{js,ts,jsx,tsx}'],
    plugins: {
      'react-hooks': pluginReactHooks
    },
    rules: {
      'react/react-in-jsx-scope': 'off',
      'react/prop-types': 'off',
      ...pluginReactHooks.configs.recommended.rules
    }
  },
  {
    ignores: ['dist/*']
  }
]

```

**Review by 7259324177 — COMMENTED** [review]
U3V30

**Comment by eps1lon** [comment]
Released in `eslint-plugin-react-hooks@5.0.0`

File: yarn.lock [hunk 3]
```diff
@@ -2130,6 +2135,18 @@
     opn "5.3.0"
     react "^16.13.1"
 
+"@eslint-community/eslint-utils@^4.2.0":
+  version "4.4.0"
+  resolved "https://registry.yarnpkg.com/@eslint-community/eslint-utils/-/eslint-utils-4.4.0.tgz#a23514e8fb9af1269d5f7788aa556798d61c6b59"
```
Comment (7259324177) [inline yarn.lock:2140]: Im thinking about this registry Api

**Review by 7259324177 — COMMENTED** [review]
Fibonacci can change to Sibonacci

**Comment by RakaDoank** [comment]
This [#issuecomment-2422831582](config) was not working for me. In my case, i have to merge the eslint-plugin-react and eslint-plugin-react-hooks in one object instead of separate the configuration

```js
export default [

	{
		ignores: [
			'**/dist/*',
			'**/node_modules/*',
		],
	},

	EslintJs.configs.recommended,

	{
		plugins: {
			'@stylistic': StylisticJs,
		},
		rules: {
			// some eslint rules here
                        // and with stylistic rules
		},
	},

	...TsEslint.configs.recommendedTypeChecked.map(conf => ({
		...conf,
		files: ['**/*.ts', '**/*.tsx', '**/*.mts'],
	})),

	{
		languageOptions: {
			parserOptions: {
				projectService: true,
				tsconfigRootDir: import.meta.dirname,
				EXPERIMENTAL_useProjectService: true,
			},
		},
	},

	{
		files: [
			'**/*.ts',
			'**/*.tsx',
			'**/*.mts',
		],
		rules: {
			'@typescript-eslint/no-unsafe-argument': 'off',
			'@typescript-eslint/no-empty-object-type': 'off',
		},
	},

	{
		files: [
			'my-monorepo-here/scripts/**/*.js',
		],
		languageOptions: {
			ecmaVersion: 6,
			sourceType: 'commonjs',
		},
	},

	{
		files: [
			'my-monorepo-here/specific-to-your-react-directory/**/*.{jsx,tsx}',
			// or you can just use '**/*.{jsx,tsx}',
		],
		plugins: {
			...ReactEslintPlugin.configs.flat['jsx-runtime'].plugins,
			'react-hooks': ReactHooksEslintPlugin,
		},
		languageOptions: {
			...ReactEslintPlugin.configs.flat['jsx-runtime'].languageOptions,
			globals: {
				...Globals.serviceworker,
				...Globals.browser,
			},
		},
		rules: {
			...ReactEslintPlugin.configs.flat.recommended.rules,
			...ReactEslintPlugin.configs.flat['jsx-runtime'].rules,
			...ReactHooksEslintPlugin.configs.recommended.rules,
		},
	},

]
```

**Comment by grubersjoe** [comment]
Another working example:

```ts
import eslint from '@eslint/js'
import pluginReact from 'eslint-plugin-react'
import pluginReactHooks from 'eslint-plugin-react-hooks'
import globals from 'globals'
import pluginTypeScript from 'typescript-eslint'

export default pluginTypeScript.config(
  eslint.configs.recommended,
  pluginTypeScript.configs.strictTypeChecked,
  pluginTypeScript.configs.stylisticTypeChecked,
  {
    rules: {
      'no-console': 'error',
    },
    languageOptions: {
      globals: globals.browser,
      parserOptions: {
        projectService: true,
        tsconfigRootDir: import.meta.dirname,
      },
    },
    settings: {
      react: {
        version: 'detect',
      },
    },
  },
  pluginReact.configs.flat.recommended,
  pluginReact.configs.flat['jsx-runtime'],
  {
    plugins: { 
      'react-hooks': pluginReactHooks,
    },
    rules: pluginReactHooks.configs.recommended.rules,
  },
  {
    // Note: there must be no other properties in this object
    ignores: ['build/', 'coverage/'],
  },
)
```

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js (hunk 1 above)
Comment (PPKk99) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:93]: eps1lon Do you have the ability to collect what you neglected?

**Pull request merged**